lib/cl/memory/bda.jl

```diff
@@ -0,0 +1,69 @@
+struct BufferDeviceMemory <: AbstractMemory
+    id::cl_mem
+    ptr::CLPtr{Cvoid}
+    bytesize::Int
+    context::Context
+end
+
+function bda_alloc(bytesize::Integer;
+        alignment::Integer = 0, device_access::Symbol = :rw, host_access::Symbol = :rw
+    )
+    bytesize == 0 && error("size 0 is not supported for BufferDeviceMemory.")
```

Comment:
Can we return a dummy object instead? Zero-size allocations are generally expected to be supported by the layers above, so if this back-end doesn't support it we should somehow spoof the object (e.g. putting in a null pointer, as we do in CUDA).

Reply:
> In CUDA.jl I've tried to confine all that logic to the high-level wrappers in src

Apparently we already do that in OpenCL, but it was still letting many edge cases slip. 
I've pushed a solution to this in the latest commit. Do let me know if you see issues with it. 

That being said, kernelabstractions and event test just segfaults on Mesa-git. I don't think they depend on BDA, so its probably a Mesa bug. 
There's 2 errors in base, but I think its because of the mentioned 'clarray views with range index' bug.